core/src/main/java/org/apache/gravitino/catalog/ModelOperationDispatcher.java

```diff
@@ -40,40 +49,114 @@ public ModelOperationDispatcher(
-    throw new UnsupportedOperationException("Not implemented");
+    NameIdentifier catalogIdent = getCatalogIdentifier(ident);
+    Map<String, String> updatedProperties = checkAndUpdateProperties(catalogIdent, properties);
+
+    Model registeredModel =
+        TreeLockUtils.doWithTreeLock(
+            NameIdentifier.of(ident.namespace().levels()),
+            LockType.WRITE,
+            () ->
+                doWithCatalog(
+                    catalogIdent,
+                    c -> c.doWithModelOps(m -> m.registerModel(ident, comment, updatedProperties)),
+                    NoSuchSchemaException.class,
+                    ModelAlreadyExistsException.class));
+
+    return EntityCombinedModel.of(registeredModel)
+        .withHiddenPropertiesSet(
+            getHiddenPropertyNames(
+                catalogIdent,
+                HasPropertyMetadata::modelPropertiesMetadata,
+                registeredModel.properties()));
   }
 
   @Override
   public boolean deleteModel(NameIdentifier ident) {
-    throw new UnsupportedOperationException("Not implemented");
+    return TreeLockUtils.doWithTreeLock(
+        NameIdentifier.of(ident.namespace().levels()),
+        LockType.WRITE,
+        () ->
+            doWithCatalog(
+                getCatalogIdentifier(ident),
+                c -> c.doWithModelOps(m -> m.deleteModel(ident)),
+                RuntimeException.class));
   }
 
   @Override
   public int[] listModelVersions(NameIdentifier ident) throws NoSuchModelException {
-    throw new UnsupportedOperationException("Not implemented");
+    return TreeLockUtils.doWithTreeLock(
+        ident,
+        LockType.READ,
+        () ->
+            doWithCatalog(
+                getCatalogIdentifier(ident),
+                c -> c.doWithModelOps(m -> m.listModelVersions(ident)),
+                NoSuchModelException.class));
   }
 
   @Override
   public ModelVersion getModelVersion(NameIdentifier ident, int version)
       throws NoSuchModelVersionException {
-    throw new UnsupportedOperationException("Not implemented");
+    return internalGetModelVersion(
+        ident,
+        () ->
+            TreeLockUtils.doWithTreeLock(
+                NameIdentifierUtil.toModelVersionIdentifier(ident, version),
+                LockType.READ,
+                () ->
+                    doWithCatalog(
+                        getCatalogIdentifier(ident),
+                        c -> c.doWithModelOps(m -> m.getModelVersion(ident, version)),
+                        NoSuchModelVersionException.class)));
   }
 
   @Override
   public ModelVersion getModelVersion(NameIdentifier ident, String alias)
       throws NoSuchModelVersionException {
-    throw new UnsupportedOperationException("Not implemented");
+    return internalGetModelVersion(
+        ident,
+        () ->
+            TreeLockUtils.doWithTreeLock(
+                NameIdentifierUtil.toModelVersionIdentifier(ident, alias),
```

Comment:
here lock the model version by model `name identifier` and `alias`, above 137 line lock the model version by model `name identifier` and `version number`.

when the alias and version point to the same model version, they will obtain two read locks, is it expected?

Reply:
I see, this is a problem, let me see how to fix it.